src/ipc/ipc4/handler.c

```diff
@@ -717,8 +717,8 @@ static int ipc4_get_large_config_module_instance(struct ipc4_message_request *ip
 					&data_offset, data);
 
 	/* set up ipc4 error code for reply data */
-	if (ret < 0)
-		ret = IPC4_MOD_INVALID_ID;
+	if (ret != 0)
```

Comment:
Not blocking the PR, but why this change? We use "ret < 0" check everywhere else in the codebase, so it looks a bit odd to use a different check in on place. Just check sof/src/ipc/ipc4/handler.c

Reply:
There was and issue with 'ret < 0'. drv->ops.get_large_config() can return positive or negative error code. For some reason, EINVAL is returned as a negative value, but e.g. IPC4_UNAVAILABLE is positive. I did not want to change error codes. 0 is always a success, so I figured out that this would be the best solution. There are other entities of 'ret != 0' in handler.c. After set_pipeline_state() and cpu_enable_core(). I do not insist on this solution, but I need a suggestion, how can we resolve this problem.

Reply:
Thanks @piotrkmx . I misunderstood the change indeed. We indeed have two error numbering schemes. SOF has traditionally used the POSIX model of negative error codes. So the root problem is that get_large_config() may return either type of error. If it's a IPC4 error (positive), it can be passed directly as a return value (which expects an IPC4 type error). Only if the function returns a negative value, we need to convert it from POSIX to IPC4. So I believe the original code is correct and this change is not needed.

Reply:
@kv2019i a slight correction: I think `.get_large_config()` is supposed to **only** return negatice POSIX errors or 0, never positive IPC4 errors. Mixing both in the same function would be bad. And no, I still don't think that even mixing both in the same subsystem is a good idea.

Reply:
> @kv2019i a slight correction: I think `.get_large_config()` is supposed to **only** return negatice POSIX errors or 0, never positive IPC4 errors. Mixing both in the same function would be bad. And no, I still don't think that even mixing both in the same subsystem is a good idea.

@lyakh IPC processing functions must implement a way to return an IPC specific status codes. I am not saying that current implementation is ideal, but we need something. The POSIX here is insufficient.